src/hooks/progress/use-update-tutorial-progress.ts

```diff
@@ -0,0 +1,186 @@
+ *
+ * On initial load, regardless of startRef or endRef visibility,
+ * if the tutorial does not yet have recorded progress,
+ * the tutorial will be marked as `visited` (0 percent).
+ *
+ * When the startRef element moves out of the viewport,
+ * the tutorial will be marked as `in_progress` (50 percent).
+ *
+ * When the endRef element moves into the viewport,
+ * the tutorial will be marked as `complete` (100 percent).
+ */
+export function useUpdateTutorialProgress({
+	tutorialId,
+	collectionId,
+}: TutorialIdCollectionId): UpdateProgressReturnInterface {
+	// We shouldn't try to update progress unless we're authenticated
+	const { isAuthenticated } = useAuthentication()
+	// We need to know if progress exists, to know whether to "create" or "update"
+	const { tutorialProgressStatus } = useTutorialProgress({
+		tutorialId,
+		collectionId,
+	})
+	// We'll use "create" or "update" progress using these `react-query` mutations
+	const tutorialProgressMutations = useTutorialProgressMutations()
+	// We use refs from `react-intersection-observer` to track progress.
+	const [startRef, startInView, startEntry] = useInView({
+		initialInView: true,
+	})
+	const [endRef, endInView, endEntry] = useInView()
+
+	/**
+	 * This effect runs often, but it seems necessary to ensure progress
+	 * gets tracked as things like the authenticated state changes.
+	 *
+	 * We bail early in many cases, and we avoid downgrading progress,
+	 * so this effect only results in meaningful mutations a small percent
+	 * of the time.
+	 */
+	useEffect(() => {
+		/**
+		 * On initial client-side navigation, this effect may be called
+		 * before startRefs and endRefs are updated. We can detect this
+		 * by comparing data-ref-id attributes we've added to the
+		 * elements startRef and endRef are attached to.
+		 */
+		const startRefId = startEntry?.target.getAttribute('data-ref-id')
+		const endRefId = endEntry?.target.getAttribute('data-ref-id')
+		const compareId = `${tutorialId}_${collectionId}`
+		const isStaticIdsMismatch =
+			startRefId !== compareId || endRefId !== compareId
+
+		console.log({
+			isStaticIdsMismatch,
+			startInView,
+			endInView,
+			startRefId,
+			endRefId,
+			compareId,
+		})
+		if (isStaticIdsMismatch) {
+			return
+		}
+
+		/**
+		 * If we're not authenticated, we bail early, there's no
+		 * point in trying to make a mutation.
+		 */
+		if (!isAuthenticated) {
+			return
+		}
+
+		/**
+		 * If we are authenticated but we don't have a result
+		 * yet from querying this tutorial's status, we also bail early.
+		 * We need that result to know whether to "create" or "update".
+		 */
+		if (typeof tutorialProgressStatus == 'undefined') {
+			return
+		}
+
+		/**
+		 * We don't want to run multiple mutations at once.
+		 * To avoid this, tutorialProgressMutations provided `isReady`,
+		 * which we respect here.
+		 */
+		const isReadyToRunMutation = tutorialProgressMutations.isReady
+		if (!isReadyToRunMutation) {
+			return
+		}
```

Comment:
I'm not sure if this check is needed since React Query already handles de-deduplicating network requests for us. It's totally cool to leave it for now for the sake of time, but could you add a ticket to our backlog to explore whether or not this is necessary? 🙏🏻

Reply:
Absolutely, added a ticket: 🎟 [investigate if react-query automatically de-duplicates requests as we expect](https://app.asana.com/0/1202097197789424/1202915360953452/f)

For context, I added this in as otherwise I was seeing very quickly repeated mutations - I think because when a mutation is called, then the result of `useTutorialProgressMutations` updates in some way, and therefore re-triggers the `useEffect` hook which has a dependency on `tutorialProgressMutations`.

Here's a screenshot just after initial load where I've removed this condition and added a `RUNNING MUTATION` log in the newly split out `updateTutorialProgress` helper:

<img width="2549" alt="CleanShot 2022-09-02 at 11 16 50@2x" src="https://user-images.githubusercontent.com/4624598/188181504-659abe94-7a9d-4e59-9a24-e801275e3e9c.png">